design/Upgrade_Design.md

```diff
@@ -0,0 +1,225 @@
+
+Note: Not critical for an upgrading mechanism to be present in phase 1.
+
+### The Wallet extension
+
+This is the component installed by end users which communicates with Obscuro nodes via RPC.
+For a good UX, Obscuro nodes must be backwards compatible and support even older versions of the Wallet extension.
+There must be warnings and mechanisms to help users upgrade to the latest version.
+
+Note: Not critical for an upgrading mechanism to be present in phase 1.
+
+## Attestation whitelisting
+
+The management contract will not support direct attestation verification, which means it has to whitelist a genesis enclave, 
+by relying on incentivising users and researchers to make sure that the attestation is the right code running on secure hardware.
+
+The genesis enclave for each new version must be published and whitelisted by the contract, together with a link to the source
+code that produces that binary image.
+
+This is the format of the event produced by the Management contract when the lifecycle is finalised.
+
+```
+ UpgradeEvent(
+   - WhitelistedGenesisAttestation // a node installed with the new version
+   - EnforcedHeight                // the rollup height from which only the new version will move forward. It can be -1 if not applicable.
+   - RotateMasterSeed              // boolean indicating whether 
+   - Attributes                    // a map of attributes that will be understood by the current version of the enclave.  
+   )
+```
+
+Note: In phase 1, the event can be emitted when the administrator of the contract calls an "approveVersion" function.
+This will give the upgrading power to the administrator, but in a transparent way.
+
+
+## Minimum viable upgrade mechanism
+
+This section describes the minimum implementation required in phase 1.
+
+Unlike most software that can get away with not thinking about upgrading during early development, TEE software doesn't have
+this luxury.
+Without a mechanism put in place in the initial release, the enclave cannot be upgraded at all because any other version will
+not be able to read the data.
+
+This is only a concern for the Obscuro enclave, as the upgrading of the Host component can be designed afterwards.
+
+### High-level tasks
+
+#### Management contract
+
+- Create the UpgradeEvent.
+- Add a function to publish metadata about a release.
+- Add a function callable only by the admin to approve one of the published releases.
+
+#### Enclave 
+
+1. Implement the UpgradeEvent authentication mechanism.
+   - Logic to understand the beacon chain canonical chain
+   - Minimal heuristics to mitigate "Weak subjectivity."
+
+2. Implement logic to verify an attestation against the approved version.
+
+3. Create an RPC endpoint to hand over secrets to an approved version, encrypted with the key.
+
+4. Logic to stop operating at the block height specified in the UpgradeEvent.
+
+
+#### Host level
+Doesn't require any special functionality in phase 1.
+
+
+### Process
```

Comment:
Perhaps it would be possible to avoid loading 2 enclaves at the same time ?

Suggestion:
Steps 1-2 : same
Step 3: When processing the UE, the CE issues some secret to the management contract that only NE can decrypt ( something using the code hash perhaps ?)
Step 4: NE reads the secret from the management contract, and process blocks in backwards mode until fork height when it enforces changes.

Reply:
> Step 3: When processing the UE, the CE issues some secret to the management contract that only NE can decrypt ( something using the code hash perhaps ?)

Was trying to avoid expensive comms through the L1. The functionality is the same in essence. 

> Step 4: NE reads the secret from the management contract, and process blocks in backwards mode until fork height when it enforces changes.

What do you mean by "process blocks in backwards mode"?

Reply:
> Was trying to avoid expensive comms through the L1. The functionality is the same in essence.

Yeah, trade-offs! I was trying to avoid having 2 enclaves running 😅 

> What do you mean by "process blocks in backwards mode"?

I mean it process blocks pre-fork.

Reply:
still not following.
Are you suggesting that the new enclave, after receiving the secret, can just sit quitely and not do anything until the block height is reached?

Reply:
No, I'm saying that the new enclave knows how to process pre and post-fork blocks. We're on the same page on this.

Overall, the difference I'm suggesting (as you picked up) is that instead of having 2 enclaves, there is only one enclave working at a time. And the CE uses the L1 to securely transmit the secret to NE.

But you're right about the cost of upgrading.. Maybe a file on disk would work ? I might be missing something but the way I see it we're just encrypting and transmitting data.

Reply:
I didn't go into the operational details on purpose.
I just listed the primitives and the constraints.
So we can play with the handover process.